src/protocol/maliciously_secure_mul.rs

```diff
@@ -0,0 +1,346 @@
+    #[must_use]
+    pub fn new(
+        ctx: &'a ProtocolContext<'a, N>,
+        record_id: RecordId,
+        security_validator: Arc<Mutex<SecurityValidator<F>>>,
+    ) -> Self {
+        Self {
+            ctx,
+            record_id,
+            security_validator,
+        }
+    }
+
+    /// Executes two parallel multiplications;
+    /// `A * B`, and `rA * B`, yielding both `AB` and `rAB`
+    /// both `AB` and `rAB` are provided to the security validator
+    ///
+    /// ## Errors
+    /// Lots of things may go wrong here, from timeouts to bad output. They will be signalled
+    /// back via the error response
+    /// ## Panics
+    /// Panics if the mutex is found to be poisoned
+    pub async fn execute(
+        self,
+        a: MaliciousReplicated<F>,
+        b: MaliciousReplicated<F>,
+    ) -> Result<MaliciousReplicated<F>, BoxError> {
+        let duplicate_ctx = self.ctx.narrow(&Step::DuplicateMultiply);
+        let (ab, rab) = try_join(
+            SecureMul::new(self.ctx, self.record_id).execute(a.x(), b.x()),
+            SecureMul::new(&duplicate_ctx, self.record_id).execute(a.rx(), b.x()),
+        )
+        .await?;
+
+        let malicious_ab = MaliciousReplicated::new(ab, rab);
+
+        self.security_validator.lock().unwrap().accumulate_macs(
+            self.ctx.narrow(&Step::RandomnessForValidation),
+            self.record_id,
+            malicious_ab,
+        );
+
+        Ok(malicious_ab)
+    }
+}
+
+#[cfg(test)]
+pub mod tests {
+    use crate::error::BoxError;
+    use crate::field::Fp31;
+    use crate::protocol::{
+        malicious::SecurityValidator, maliciously_secure_mul::MaliciouslySecureMul, QueryId,
+        RecordId,
+    };
+    use crate::secret_sharing::{MaliciousReplicated, Replicated};
+    use crate::test_fixture::{
+        logging, make_contexts, make_world, share, validate_and_reconstruct, TestWorld,
+    };
+
+    use futures::future::{try_join, try_join_all};
+    use proptest::prelude::Rng;
+    use std::sync::{Arc, Mutex};
+
+    /// This is the simplest arithmetic circuit that allows us to test all of the pieces of this validator
+    /// A -
+    ///     \
+    ///      Mult_Gate -> A*B
+    ///     /
+    /// B -
+    ///
+    /// This circuit has two inputs, A and B. These two inputs are multiplied together. That's it.
+    ///
+    /// To acheive malicious security, the entire circuit must be run twice, once with the original inputs,
+    /// and once with all the inputs times a random, secret-shared value `r`. Two information theoretic MACs
+    /// are updated; once for each input, and once for each multiplication. At the end of the circuit, these
+    /// MACs are compared. If any helper deviated from the protocol, chances are that the MACs will not match up.
+    /// There is a small chance of failure which is `2 / |F|`, where `|F|` is the cardinality of the prime field.
+    #[tokio::test]
+    async fn test_simplest_circuit() -> Result<(), BoxError> {
```

Comment:
Am I looking at code duplication?  Is there anything you can factor out between these two very similar tests?